File: fe/fe-core/src/main/java/com/starrocks/catalog/combinator/AggStateDesc.java
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.starrocks.catalog.combinator;

import com.google.api.client.util.Lists;
import com.google.common.base.Objects;
import com.google.common.base.Preconditions;
import com.google.gson.annotations.SerializedName;
import com.starrocks.analysis.FunctionParams;
import com.starrocks.catalog.AggregateFunction;
import com.starrocks.catalog.Function;
import com.starrocks.catalog.FunctionSet;
import com.starrocks.catalog.Type;
import com.starrocks.common.AnalysisException;
import com.starrocks.qe.ConnectContext;
import com.starrocks.sql.analyzer.FunctionAnalyzer;
import com.starrocks.sql.parser.NodePosition;
import com.starrocks.thrift.TAggStateDesc;
import com.starrocks.thrift.TFunctionVersion;
import com.starrocks.thrift.TTypeDesc;
import com.starrocks.thrift.TTypeNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * AggStateDesc is used to get the aggregate function which's used for common aggregate function state. 
 * Aggregate function's state needs to find the associated aggregate function, to get the aggregate function, 
 * we need to know the function name, return type, argument types and result nullable.
 */
public class AggStateDesc {
    // aggregate function's name
    @SerializedName(value = "functionName")
    private final String functionName;
    // aggregate function's argument types
    @SerializedName(value = "argTypes")
    private final List<Type> argTypes;
    // aggregate function's return type
    @SerializedName(value = "returnType")
    private final Type returnType;
    // aggregate function's result nullable
    @SerializedName(value = "resultNullable")
    private final Boolean resultNullable;

    public AggStateDesc(AggregateFunction aggFunc) {
        this(aggFunc.functionName(), aggFunc.getReturnType(), Arrays.asList(aggFunc.getArgs()));
    }

    public AggStateDesc(String functionName,
                        Type returnType,
                        List<Type> argTypes) {
        Preconditions.checkNotNull(functionName, "functionName should not be null");
        Preconditions.checkNotNull(returnType, "returnType should not be null");
        Preconditions.checkNotNull(argTypes, "argTypes should not be null");
        this.functionName = functionName;
        this.returnType = returnType;
        this.argTypes = argTypes;
        this.resultNullable = isAggFuncResultNullable(functionName);
    }

    private boolean isAggFuncResultNullable(String functionName) {
        // To be more compatible, always set result nullable to true here. This may decrease the performance of runtime
        // but can be more compatible with different aggregate functions and inputs.
        // this.resultNullable = !FunctionSet.alwaysReturnNonNullableFunctions.contains(functionName);
        if (FunctionSet.COUNT.equalsIgnoreCase(functionName)) {
            return false;
        } else {
            return true;
        }
    }

    public List<Type> getArgTypes() {
        return argTypes;
    }

    public Type getReturnType() {
        return returnType;
    }

    public Boolean getResultNullable() {
        return resultNullable;
    }

    public String getFunctionName() {
        return functionName;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(functionName, argTypes, returnType, resultNullable);
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof AggStateDesc)) {
            return false;
        }
        AggStateDesc other = (AggStateDesc) o;
        // result nullable
        if (resultNullable != other.resultNullable) {
            return false;
        }
        // func name
        if (!functionName.equalsIgnoreCase(other.functionName)) {
            return false;
        }
        // return type
        if (!returnType.equals(other.returnType)) {
            return false;
        }
        // arg types
        int subTypeNumber = argTypes.size();
        if (subTypeNumber != other.argTypes.size()) {
            return false;
        }
        for (int i = 0; i < subTypeNumber; i++) {
            if (!argTypes.get(i).equals(other.argTypes.get(i))) {
                return false;
            }
        }
        return true;
    }

    public TAggStateDesc toThrift() {
        // wrapper extra data type
        TAggStateDesc tAggStateDesc = new TAggStateDesc();
        tAggStateDesc.setAgg_func_name(Function.rectifyFunctionName(functionName));
        for (Type argType : argTypes) {
            TTypeDesc tTypeDesc = new TTypeDesc();
            tTypeDesc.setTypes(new ArrayList<TTypeNode>());
            argType.toThrift(tTypeDesc);
            tAggStateDesc.addToArg_types(tTypeDesc);
        }
        tAggStateDesc.setResult_nullable(resultNullable);
        tAggStateDesc.setFunc_version(TFunctionVersion.RUNTIME_FILTER_SERIALIZE_VERSION_3.getValue());

        // ret type
        TTypeDesc tTypeDesc = new TTypeDesc();
        tTypeDesc.setTypes(new ArrayList<TTypeNode>());
        returnType.toThrift(tTypeDesc);
        tAggStateDesc.setRet_type(tTypeDesc);
        Preconditions.checkState(!tTypeDesc.types.isEmpty());

        return tAggStateDesc;
    }

    public AggStateDesc clone() {
        return new AggStateDesc(functionName, returnType, Lists.newArrayList(argTypes));
    }

    /**
     * @return associated analyzed aggregate function
     * @throws AnalysisException: when the aggregate function is not found or not an aggregate function
     */
    public AggregateFunction getAggregateFunction() throws AnalysisException {
        FunctionParams params = new FunctionParams(false, Lists.newArrayList());
        Type[] argumentTypes = argTypes.toArray(Type[]::new);
        Boolean[] isArgumentConstants = argTypes.stream().map(x -> false).toArray(Boolean[]::new);
        Function result = FunctionAnalyzer.getAnalyzedAggregateFunction(ConnectContext.get(),
                this.functionName, params, argumentTypes, isArgumentConstants, NodePosition.ZERO);
        if (result == null) {
            throw new AnalysisException(String.format("AggStateType function %s with input %s not found", functionName,
                    argTypes));
        }
        if (!(result instanceof AggregateFunction)) {
            throw new AnalysisException(String.format("AggStateType function %s with input %s found but not an aggregate " +
                    "function", functionName, argTypes));
        }
        return (AggregateFunction) result;
    }

    @Override
    public String toString() {
        return toSql();
    }

    public String toSql() {
        StringBuilder sb = new StringBuilder();
        sb.append(functionName).append("(");
        for (int i = 0; i < argTypes.size(); i++) {
            if (i > 0) {
                sb.append(", ");
            }
            sb.append(argTypes.get(i).toSql());
        }
        sb.append(")");
        return sb.toString();
    }
}

File: be/src/runtime/agg_state_desc.cpp
// Copyright 2021-present StarRocks, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "runtime/agg_state_desc.h"

#include <memory>

#include "exprs/agg/aggregate.h"
#include "exprs/agg/aggregate_factory.h"

namespace starrocks {

// Create a new AggStateDesc from a thrift TTypeDesc.
AggStateDesc AggStateDesc::from_thrift(const TAggStateDesc& desc) {
    VLOG(2) << "TAggStateDesc:" << apache::thrift::ThriftDebugString(desc);
    std::string agg_func_name = desc.agg_func_name;
    // return type
    auto return_type = TypeDescriptor::from_thrift(desc.ret_type);
    // arg types
    std::vector<TypeDescriptor> arg_types;
    for (auto& arg_type : desc.arg_types) {
        arg_types.emplace_back(TypeDescriptor::from_thrift(arg_type));
    }
    bool result_nullable = desc.result_nullable;
    int func_version = desc.func_version;
    return AggStateDesc{agg_func_name, std::move(return_type), std::move(arg_types), result_nullable, func_version};
}

// Transform this AggStateDesc to a thrift TTypeDesc.
void AggStateDesc::to_thrift(TAggStateDesc* t) {
    t->agg_func_name = _func_name;
    t->result_nullable = _is_result_nullable;
    t->func_version = _func_version;
    // return type
    t->ret_type = _return_type.to_thrift();
    // arg types
    for (auto& arg_type : _arg_types) {
        t->arg_types.push_back(arg_type.to_thrift());
    }
}

AggStateDesc AggStateDesc::from_protobuf(const AggStateDescPB& desc) {
    auto& agg_func_name = desc.agg_func_name();
    bool is_result_nullable = desc.is_result_nullable();
    int func_version = desc.func_version();
    std::vector<TypeDescriptor> arg_types;
    // arg types
    for (auto& arg_type : desc.arg_types()) {
        arg_types.emplace_back(TypeDescriptor::from_protobuf(arg_type));
    }
    // ret type
    auto ret_type = TypeDescriptor::from_protobuf(desc.ret_type());
    return AggStateDesc{agg_func_name, std::move(ret_type), std::move(arg_types), is_result_nullable, func_version};
}

void AggStateDesc::to_protobuf(AggStateDescPB* desc) {
    desc->set_agg_func_name(this->get_func_name());
    desc->set_is_result_nullable(this->is_result_nullable());
    desc->set_func_version(this->get_func_version());
    // arg types
    for (auto& arg_type : this->get_arg_types()) {
        auto* arg_type_pb = desc->add_arg_types();
        *arg_type_pb = arg_type.to_protobuf();
    }
    // ret type
    auto ret_type_desc = this->get_return_type();
    auto* ret_type_pb = desc->mutable_ret_type();
    *ret_type_pb = ret_type_desc.to_protobuf();
}

void AggStateDesc::thrift_to_protobuf(const TAggStateDesc& desc, AggStateDescPB* pb) {
    pb->set_agg_func_name(desc.agg_func_name);
    pb->set_is_result_nullable(desc.result_nullable);
    pb->set_func_version(desc.func_version);
    // arg types
    for (auto& arg_type : desc.arg_types) {
        auto arg_type_desc = TypeDescriptor::from_thrift(arg_type);
        auto* arg_type_pb = pb->add_arg_types();
        *arg_type_pb = arg_type_desc.to_protobuf();
    }
    // ret type
    auto ret_type_desc = TypeDescriptor::from_thrift(desc.ret_type);
    auto* ret_type_pb = pb->mutable_ret_type();
    *ret_type_pb = ret_type_desc.to_protobuf();
}

std::string AggStateDesc::debug_string() const {
    std::stringstream ss;
    ss << "[" << _func_name << ", args:<";
    for (size_t i = 0; i < _arg_types.size(); i++) {
        if (i != _arg_types.size() - 1) {
            ss << _arg_types[i] << ", ";
        } else {
            ss << _arg_types[i] << ">";
        }
    }
    ss << ", ret:" << _return_type << ", result_nullable:" << _is_result_nullable << ", func_version:" << _func_version
       << "]";
    return ss.str();
}

const AggregateFunction* AggStateDesc::get_agg_state_func(AggStateDesc* agg_state_desc) {
    DCHECK(agg_state_desc);
    auto* agg_function = get_aggregate_function(agg_state_desc->get_func_name(), agg_state_desc->get_return_type(),
                                                agg_state_desc->get_arg_types(), agg_state_desc->is_result_nullable(),
                                                TFunctionBinaryType::BUILTIN, agg_state_desc->get_func_version());
    if (agg_function == nullptr) {
        LOG(WARNING) << "Failed to get aggregate function for " << agg_state_desc->debug_string();
    }
    return agg_function;
}

} // namespace starrocks